processor/intervalprocessor/processor.go

```diff
@@ -52,5 +96,182 @@ func (p *Processor) Capabilities() consumer.Capabilities {
 }
 
 func (p *Processor) ConsumeMetrics(ctx context.Context, md pmetric.Metrics) error {
-	return p.nextConsumer.ConsumeMetrics(ctx, md)
+	var errs error
+
+	p.stateLock.Lock()
+	defer p.stateLock.Unlock()
+
+	md.ResourceMetrics().RemoveIf(func(rm pmetric.ResourceMetrics) bool {
+		rm.ScopeMetrics().RemoveIf(func(sm pmetric.ScopeMetrics) bool {
+			sm.Metrics().RemoveIf(func(m pmetric.Metric) bool {
+				switch m.Type() {
+				case pmetric.MetricTypeGauge, pmetric.MetricTypeSummary:
+					return false
+				case pmetric.MetricTypeSum:
+					// Check if we care about this value
+					sum := m.Sum()
+
+					if !sum.IsMonotonic() {
+						return false
+					}
+
+					if sum.AggregationTemporality() != pmetric.AggregationTemporalityCumulative {
+						return false
+					}
+
+					mClone, metricID := p.getOrCloneMetric(rm, sm, m)
+					cloneSum := mClone.Sum()
+
+					aggregateDataPoints(sum.DataPoints(), cloneSum.DataPoints(), metricID, p.numberLookup)
+					return true
+				case pmetric.MetricTypeHistogram:
+					histogram := m.Histogram()
+
+					if histogram.AggregationTemporality() != pmetric.AggregationTemporalityCumulative {
+						return false
+					}
+
+					mClone, metricID := p.getOrCloneMetric(rm, sm, m)
+					cloneHistogram := mClone.Histogram()
+
+					aggregateDataPoints(histogram.DataPoints(), cloneHistogram.DataPoints(), metricID, p.histogramLookup)
+					return true
+				case pmetric.MetricTypeExponentialHistogram:
+					expHistogram := m.ExponentialHistogram()
+
+					if expHistogram.AggregationTemporality() != pmetric.AggregationTemporalityCumulative {
+						return false
+					}
+
+					mClone, metricID := p.getOrCloneMetric(rm, sm, m)
+					cloneExpHistogram := mClone.ExponentialHistogram()
+
+					aggregateDataPoints(expHistogram.DataPoints(), cloneExpHistogram.DataPoints(), metricID, p.expHistogramLookup)
+					return true
+				default:
+					errs = errors.Join(fmt.Errorf("invalid MetricType %d", m.Type()))
+					return false
+				}
+			})
+			return sm.Metrics().Len() == 0
+		})
+		return rm.ScopeMetrics().Len() == 0
+	})
+
+	if err := p.nextConsumer.ConsumeMetrics(ctx, md); err != nil {
+		errs = errors.Join(errs, err)
+	}
+
+	return errs
+}
+
+func aggregateDataPoints[DPS metrics.DataPointSlice[DP], DP metrics.DataPoint[DP]](dataPoints DPS, mCloneDataPoints DPS, metricID identity.Metric, dpLookup map[identity.Stream]DP) {
+	for i := 0; i < dataPoints.Len(); i++ {
+		dp := dataPoints.At(i)
+
+		streamID := identity.OfStream(metricID, dp)
+		existingDP, ok := dpLookup[streamID]
+		if !ok {
+			dpClone := mCloneDataPoints.AppendEmpty()
+			dp.CopyTo(dpClone)
+			dpLookup[streamID] = dpClone
+			continue
+		}
+
+		// Check if the datapoint is newer
+		if dp.Timestamp() > existingDP.Timestamp() {
+			dp.CopyTo(existingDP)
+			continue
+		}
+
+		// Otherwise, we leave existing as-is
+	}
+}
+
+func (p *Processor) exportMetrics() {
+	md := func() pmetric.Metrics {
```

Comment:
why the closure?

Reply:
So the lock only lasts for the fetch, and not for the duration of ConsumeMetrics. Since defer is on function exit, not scope exit.